### PR TITLE
[SpellAura] Dose flame touched aura on flame sear

### PR DESCRIPTION
**Pull request by Patrick-Edouard**
The Eredar twins from Sunwell Plateau.

Flame sear dots should add stacks of Flame Touched debuff.

In the current state of mangos, it only applies an initial flame touched when it lands, but not on DoT ticks. (Can provide youtube video if requiered).

The flame sear dot lasts for 6 second ticking every 500ms. But on youtube video we clearly see it stack one flame touched per second, for a total of 6 stacks. I assume there is a hiden cooldown mechanic that should maybe be managed with the dark flame auras.

### CODE REVIEW

File: src/game/SpellAuras.cpp [hunk 1]
```diff
@@ -5724,6 +5724,19 @@ void Aura::PeriodicTick()
                         }
                         break;
                     }
+                    case 46771: {                               // Flame Sear
+                        if (target->GetTypeId() == TYPEID_PLAYER) {
```
Comment (killerwife) [inline src/game/SpellAuras.cpp]: Please use C++ brackets styling, not Java :P

File: src/game/SpellAuras.cpp [hunk 2]
```diff
@@ -5724,6 +5724,19 @@ void Aura::PeriodicTick()
...
+                            if (!target->HasAura(45343) &&
+                                !target->HasAura(47300)) {          // No Dark Flame applied. There are two auras
+                                if (m_periodicTick % 4 == 1) {      // Only doze the aura every 2 ticks (should maybe be one second cooldown)
+                                    target->CastSpell(target, 45348,
```
Comment (killerwife) [inline src/game/SpellAuras.cpp]: I believe you should apply different debuffs on 45343 presence, and different on 47300. Right now you apply only fire one.

**Comment by killerwife** [comment]
Also on a general note, this code should be handled in UnitAuraProcHandler.cpp not PeriodicTick() function.

File: src/game/SpellAuras.cpp (hunk 2 above)
Comment (Patrick-Edouard) [inline src/game/SpellAuras.cpp]: The boss mecanic is the following:

- One fire lady does fire damage and applies fire debuff when it ticks
- One shadow lady does shadow damage and applies shadow debuff when it ticks
- When someone has a fire or a shadow debuff and take damage from the opposite school, he gains 3 seconds of immunity for all debuff applications

I assume one aura is triggered when switching from fire to shadow and the other the other why around.

This test seems legit to me

Comment (killerwife) [inline src/game/SpellAuras.cpp]: The check currently does "if no dark flame aura and if no fire flame aura do stuff", afaik u want "if dark flame aura procs do stuff, if currently has flame touched, run immunity" "if fire flame aura procs, do stuff, if currently has shadow touched, run immunity" or am I wrong?

Comment (Patrick-Edouard) [inline src/game/SpellAuras.cpp]: What I want is :

If flame sear (id: 46771) dot is ticking
And if there is no dark flame aura (id: 45343 and 47300) (3 second imunity to fire auras)
Then do stuff (increase fire aura stacks)

**Comment by Qingle** [comment]
Another additional comment, shouldn't 'target' be 'unitTarget'?

**Comment by Patrick-Edouard** [comment]
@Qingle I have no clue about that; but in this function everyone seem to use 'target' that is defined at its beginning :

`Unit* target = GetTarget();`

**Comment by Qingle** [comment]
@Patrick-Edouard Yes apologies, I'm too used to doing Dummy Auras ^^

**Comment by Patrick-Edouard** [comment]
It seems the spell 45342 also have a wrong behavior : applies a stack of flame touched on landing, then nothing on ticks. 

CF retail video : https://youtu.be/AKdBJFzvJcY?t=1m30s

**Comment by Qingle** [comment]
Just looked at the spell, it applies a periodic damage effect, a confuse and also target everyone around the caster and applies damage to them.

Have you got QSpellWork and a Spell.dbc file to work with?

**Comment by Patrick-Edouard** [comment]
@Qingle Yo,

No I haven't started to look in this aspect of wowemu yet...

To be clear, what was triggering me is that flame touched debuff (id: 45348) isn't stacking properly on fire spells ticks (for flame sear id:46771 and conflagration id:45342 ), instead it just stack on spell landing.

So I think the root of the evil is that the current flame touched / shadow touched / dark flame is handled in the boss script here:

https://github.com/cmangos/mangos-tbc/blob/master/src/scriptdev2/scripts/eastern_kingdoms/sunwell_plateau/boss_eredar_twins.cpp#L192

While it should be spell effect? So reworking it would mean reworking all the encounter spells (flame sear, conflagration, blaze, shadow nova, shadow blade, flame touched, shadow touched, dark flame 1, dark flame  2). And delete this SpellHitTarget method.

It seems a bit too big for me ^^

**Comment by Qingle** [comment]
Then I feel as @killerwife said, this should be done in UnitAuraProcHandler.cpp

**Comment by Patrick-Edouard** [comment]
@Qingle @killerwife 

I've tried to play with `spell_proc_event` and add some of twins spell into it in order to make it trigger `Unit::HandleProcTriggerSpellAuraProc` or other functions `UnitAuraProcHandler.cpp` but without any result..

Some of the spells that I can trace from `HandleProcTriggerSpellAuraProc` are not even in this table. So I don't even know if I'm looking in the right direction.

It seems that trinity is using a table named `spell_link_spell`to handle this mecanic, but I can't find any equivalent in mangos.

I'm most likely going spend a few hours this week end, if I don't get forward I'll give up and just make an issue ^^

**Comment by killerwife** [comment]
Ok, ill try to look this evening.

**Comment by killerwife** [comment]
Apply this patch https://gist.github.com/killerwife/e0665b3b12435abef543562d8bf16bb0 after that it enters 

SpellAuraProcResult Unit::HandleProcTriggerSpellAuraProc(Unit* pVictim, uint32 damage, Aura* triggeredByAura, SpellEntry const* procSpell, uint32 procFlags, uint32 procEx, uint32 cooldown)

and you can script away. The proc should do all, immunities, flame touched/shadow touched. Also since there are two Dark Flame auras, I think one should be named "Normal Flame Aura" and should work for fire spells.

**Comment by Patrick-Edouard** [comment]
Ok, I made some changes to handle those procs.

It's still not perfect and some part are a bit hacky. One of the main concern is about the number of stacks. On Flame Sear there should be 6 stack of flame touched and I got 7 with this:
CF video : https://www.twitch.tv/vooml/v/116256396

**Comment by Patrick-Edouard** [comment]
Second concern is about this: https://github.com/cmangos/mangos-tbc/pull/175/files#diff-a35c8aeb0556a8ce0acb08c987d3d23bR1744

The remove aura should be handled in the dark flame proc?

**Comment by cyberium** [comment]
@killerwife sound like abandonned/ not correct? Close this if any.

**Comment by Patrick-Edouard** [comment]
@cyberium well it's the blizzlike behavior. The server I was playing on merged this and since I had no feedback here, I had no incentive to keep posting.

**Comment by killerwife** [comment]
@cyberium no, all that needs to be done is that it needs to be rebased properly, so that the changes are visible. This PR actually seems to be doing it properly, its just hard to see the changes.

**Comment by cyberium** [comment]
@killerwife cleaned here -> https://gist.github.com/cyberium/89815c9630c21e9aaab90f06d1141a70
You have no excuse now :P

**Comment by killerwife** [comment]
@cyberium if you fix up formatting a bit, then its indeed good to go. I remember helping @Patrick-Edouard a bit on IRC but I think due to school I forgot about it. My bad.

**Comment by Patrick-Edouard** [comment]
@killerwife just a head up. Seems cyberium made the so called clean up. Where you expecting me to report it to the pull request or something?

**Comment by killerwife** [comment]
I saw the cleaned up code. Someone, either you, or someone else has to punch it into git and commit it to the repo. I am spreading myself so thin lately that I don't keep up with everything.

**Pull request closed without merging**